immortal-cravings.lua

```diff
@@ -0,0 +1,231 @@
+    enabled = persisted_data.enabled or false
+end
+
+DrinkAlcohol = df.need_type['DrinkAlcohol']
+EatGoodMeal = df.need_type['EatGoodMeal']
+
+---@type integer[]
+watched = {}
+
+threshold = -9000
+
+---unit loop: check for idle watched units and create eat/drink jobs for them
+local function unit_loop()
+    -- print(('immortal-cravings: running unit loop (%d watched units)'):format(#watched))
+    ---@type integer[]
+    local kept = {}
+    for _, unit_id in ipairs(watched) do
+        local unit = df.unit.find(unit_id)
+        if unit and not (unit.flags1.caged or unit.flags1.chained) then
+            if not idle.unitIsAvailable(unit) then
+                table.insert(kept, unit.id)
+            else
+                --
+                for _, need in ipairs(unit.status.current_soul.personality.needs) do
+                    if need.id == DrinkAlcohol and need.focus_level < threshold then
+                        goDrink(unit)
+                        goto next_unit
+                    elseif need.id == EatGoodMeal and need.focus_level < threshold then
+                        goEat(unit)
+                        goto next_unit
+                    end
+                end
+            end
+        else
+            -- print('immortal-cravings: unit gone or caged')
+        end
+        ::next_unit::
+    end
+    watched = kept
+    if #watched == 0 then
+        -- print('immortal-cravings: no more watched units, cancelling unit loop')
+        repeatutil.cancel(GLOBAL_KEY .. '-unit')
+    end
+end
+
+---main loop: look for citizens with personality needs for food/drink but w/o physiological need
+local function main_loop()
+    print('immortal-cravings watching:')
+    watched = {}
+    for _, unit in ipairs(dfhack.units.getCitizens()) do
+        if unit.curse.add_tags1.NO_DRINK or unit.curse.add_tags1.NO_EAT then
+            for _, need in ipairs(unit.status.current_soul.personality.needs) do
+                if need.id == DrinkAlcohol and need.focus_level < threshold or
+                    need.id == EatGoodMeal and need.focus_level < threshold
+                then
+                    table.insert(watched, unit.id)
+                    print('  '..dfhack.df2console(dfhack.TranslateName(dfhack.units.getVisibleName(unit))))
+                    goto next_unit
+                end
+            end
+        end
+        ::next_unit::
+    end
+
+    if #watched > 0 then
+        repeatutil.scheduleUnlessAlreadyScheduled(GLOBAL_KEY..'-unit', 59, 'ticks', unit_loop)
+    end
+end
+
+local function start()
+    if enabled then
+        repeatutil.scheduleUnlessAlreadyScheduled(GLOBAL_KEY..'-main', 4003, 'ticks', main_loop)
+    end
+end
+
+local function stop()
+    repeatutil.cancel(GLOBAL_KEY..'-main')
+    repeatutil.cancel(GLOBAL_KEY..'-unit')
+end
+
+
+
+-- script action
+
+--- Handles automatic loading
+dfhack.onStateChange[GLOBAL_KEY] = function(sc)
+    if sc == SC_MAP_UNLOADED then
+        enabled = false
```

Comment:
maybe add a comment here explaining that repeat-util will cancel the scheduled callbacks for you